build: fix incorrect initramfs (bzip2, gzip) compression

Spudz76 opened the pull request:
> build: fix incorrect initramfs (bzip2, gzip) compression
> 
> Requires: tools/bzip2, tools/libdeflate
> 
> fix consistency of executables so all of them explicitly use
> `$(STAGING_DIR_HOST)/bin`, and not system-installed ones from the usual
> environment `PATH`; this affects these options:
> * `CONFIG_KERNEL_INITRAMFS_COMPRESSION_BZIP2`
> * `CONFIG_KERNEL_INITRAMFS_COMPRESSION_GZIP`
> 
> those may have worked in the past but only via side effect of having the
> binaries on the host system (and whatever unpredictable version or
> patchset those might be)
>
> use newer libdeflate-gzip for the GZIP support
>
> Fixes: 330bd380e8b6 ("image: allow building FIT and uImage with ramdisk")

This is split from [another initramfs PR](https://github.com/openwrt/openwrt/pull/11381) because this part fixes a past commit.

Also includes a small new fix for missing and incorrect prereq checks, related to gzip but unrelated to initramfs.

Spudz76 committed:
tools/bzip2: add `bzip2` binaries

`bzip2` is the standard executable for bzip2 compression

this includes development includes and both static and shared libs
(libbz2) which can be used by other packages

the initramfs generator offers the BZIP2 option but there was no
executable to support it, and worked only via side effect of having a
system-installed version of bzip2, which could be less predictable

Signed-off-by: Tony Butler <spudz76@gmail.com>

Spudz76 commented:
Merged `tools/libbz2` into `tools/bzip2` per direction in another PR, and improved the upstream source Makefile situation completely (unified split Makefiles, and added ala-carte features, and completed `install` targets)

dangowrt commented:
`bzip2` is a build prerequisite, and we do check for it being present:
https://github.com/openwrt/openwrt/blob/master/include/prereq-build.mk#L156

However, we also heavily rely on `gzip` being present and don't check for that.

Not sure where to go from there...

Spudz76 commented:
Well obviously check for `gzip` #11433 since it's used for unpacking downloads, and other tasks.

But also using OS-provided compressor tools for initramfs or squashfs is not a good idea for reproducible builds.  So a bundled fully controlled binary of known provenance is the only option.  And a majority already used bundled binaries so this just makes them all consistent.

Ansuel commented on tools/Makefile:
```diff
@@ -67,6 +67,7 @@ tools-y += sstrip
 tools-y += zip
 tools-y += zlib
 tools-y += zstd
+tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS),y) += liblzo
```
why?

Spudz76 replied:
Accidental leftover from original PR where lz4/lzo were involved.  Cleaning up while implementing these other suggestions.

Spudz76 replied:
Removed.

Ansuel commented:
@Spudz76 can you split the last commit in 2? One for gzip and the other for bzip2 ?

Also for the bzip2 commit, I would only add the required makefile and patches and make the tools changes in the related commit tweaking the bzip2 part for kernel-defaults

Ansuel commented:
Also where the patch comes from?

Spudz76 commented:
Patch came from myself (?) as I also wrote the PR referenced in the initial post but it was even more of a combined blob/mess.  Nothing really uses the initramfs since squashfs/UBI are fairly universally supported (and don't waste kernel-partition space).

Generally I hack and hack and hack (messily) on my own dev-tree and then later decide some things could be useful for other people.  There are all sorts of hacks that aren't "public safe" or of questionable utility for anyone else and sometimes I don't split them out quite perfectly, or I get lost as to which feature goes with which mods.  And then when one gets merged I have to do fairly messy rebasing of my own dev-tree and sometimes leftovers linger.

Ansuel commented:
@Spudz76 sorry was probably confusing... referring to the bzip2 patch... there is that second patch that reorganize the makefile and i can't understand why?

Spudz76 commented:
Well it's been so long, I don't actually remember.  The `tools/bzip2` got merged separately since this began, but apparently I didn't like how it ended up for some reason, so now there are adjustments here.  And I think the bot complained about patches needing rebase so I did that too.

Spudz76 commented:
Unless you mean the semi-unrelated PR for Makefile reorg.  I can never quite tell which thing will be merged first or what should be separate (per component) or what should be bundled (per goal).  I tend towards grouping by goal but some things are used by multiple goals (like libbz2 is used by the squashfs version of this goal, and was separate, but someone told me to merge it with the bzip2 PR) and then I attempt to get those merged in the correct order.

Mainly it's the too many things in-flight and interdepending and then keeping them all rebased to still work with other changes.

Ansuel commented:
@Spudz76 i'm referring to 020 and 021 for bzip2 

They are backports? Especially why 021 is needed?

Spudz76 committed:
build: fix incorrect initramfs bzip2 compression

Requires: tools/bzip2

fix consistency of executable to use `$(STAGING_DIR_HOST)/bin/bzip2`, and
not system-installed ones from the usual environment `PATH`;
this affects option `CONFIG_KERNEL_INITRAMFS_COMPRESSION_BZIP2`

this may have worked in the past but only via side effect of having the
binaries on the host system (and whatever unpredictable version or
patchset those might be)

Fixes: 330bd380e8b6 ("image: allow building FIT and uImage with ramdisk")
Signed-off-by: Tony Butler <spudz76@gmail.com>

Spudz76 committed:
build: fix incorrect initramfs gzip compression

Requires: tools/libdeflate

fix consistency of executable to use
`$(STAGING_DIR_HOST)/bin/libdeflate-gzip`, and not system-installed ones
from the usual environment `PATH`;
this affects option `CONFIG_KERNEL_INITRAMFS_COMPRESSION_GZIP`

this may have worked in the past but only via side effect of having the
binaries on the host system (and whatever unpredictable version or
patchset those might be), and did not use the improved but totally
compatible libdeflate-gzip

Fixes: 330bd380e8b6 ("image: allow building FIT and uImage with ramdisk")
Signed-off-by: Tony Butler <spudz76@gmail.com>

Spudz76 commented:
patch `020` is needed for compile on BSD stuff (including OSX).  Not sure where I found out it's needed.

patch `021` makes it so normal generic settings from the build system work properly instead of having custom (imho duplicated) procedures in the `tools/bzip2/Makefile` for compile settings and install.

Spudz76 commented:
Also, split up into more commits, I think something like what was asked.

Ansuel commented on tools/Makefile:
```diff
@@ -64,8 +67,8 @@ tools-y += sstrip
 tools-y += zip
 tools-y += zlib
 tools-y += zstd
-tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS),y) += liblzo
```
@Spudz76 it seems the rebase did some magic and the removal of this cause problem with sdk compilation (or compilation problem by using lzo compression... Can you by chance check this? I'm confused on why this actually works (considering liblzo is a dependency but in theory falls in tools-n list)

Spudz76 replied:
If anything else pulls in `lzop` then lzop pulls in liblzo.  Squashfs4 code only needs liblzo, initramfs uses lzop.
